File: September 2022/September_04.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        TreeMap<Integer, List<Integer>> map = new TreeMap<>();
        Queue<Pair> q = new ArrayDeque<>();
        q.add(new Pair(root, 0));
        while(!q.isEmpty()){
            int size = q.size();
            Map<Integer, List<Integer>> tmp = new HashMap<>();
            while(size-- > 0){
                Pair rem = q.remove();
                if(!tmp.containsKey(rem.angle))
                    tmp.put(rem.angle, new ArrayList<>());
                tmp.get(rem.angle).add(rem.node.val);
                if(rem.node.left != null)
                    q.add(new Pair(rem.node.left, rem.angle - 1));
                if(rem.node.right != null)
                    q.add(new Pair(rem.node.right, rem.angle + 1));
            }
            for(int k : tmp.keySet()){
                Collections.sort(tmp.get(k));
                if(!map.containsKey(k))
                    map.put(k, new ArrayList<>());
                map.get(k).addAll(tmp.get(k));
            }
        }
        List<List<Integer>> list = new ArrayList<>();
        for(int k : map.keySet())
            list.add(map.get(k));
        return list;
    }
    class Pair{
        TreeNode node;
        int angle;
        Pair(TreeNode node, int angle){
            this.node = node;
            this.angle = angle;
        }
    }
}

File: September 2022/September_04.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        map<int, vector<int> > mp;
        queue<pair<TreeNode*, int> > q;
        q.push({root, 0});
        while(!q.empty()){
            int size = q.size();
            unordered_map<int, vector<int> > tmp;
            while(size-- > 0){
                pair<TreeNode*, int> rem = q.front();
                q.pop();
                if(tmp.find(rem.second) == tmp.end()){
                    vector<int> v;
                    tmp[rem.second] = v;
                }
                tmp[rem.second].push_back(rem.first->val);
                if(rem.first->left != nullptr)
                    q.push({rem.first->left, rem.second - 1});
                if(rem.first->right != nullptr)
                    q.push({rem.first->right, rem.second + 1});
            }
            for(auto &k : tmp){
                sort(k.second.begin(), k.second.end());
                if(mp.find(k.first) == mp.end()){
                    vector<int> v;
                    mp[k.first] = v;
                }
                for(int i : k.second)
                    mp[k.first].push_back(i);
            }
        }
        vector<vector<int>> vec;
        for(auto &k : mp)
            vec.push_back(k.second);
        return vec;
    }
};
